polkadot/consensus/src/lib.rs

```diff
@@ -495,133 +252,211 @@ impl<C: PolkadotApi, N: Network> bft::ProposerFactory for ProposerFactory<C, N>
+			client: self.client.clone(),
+			collators: self.collators.clone(),
+			delay: timeout.shared(),
+			dynamic_inclusion,
+			local_duty,
+			local_key: sign_with,
 			parent_hash,
-			parent_number: parent_header.number,
 			parent_id: checked_id,
-			local_key: sign_with,
-			client: self.client.clone(),
+			parent_number: parent_header.number,
+			random_seed,
+			router,
+			table,
+			timer: self.timer.clone(),
 			transaction_pool: self.transaction_pool.clone(),
-			delay: timeout.shared(),
-			_table: table,
-			_router: router,
 		})
 	}
 }
 
-fn current_timestamp() -> Timestamp {
-	use std::time;
-
-	time::SystemTime::now().duration_since(time::UNIX_EPOCH)
-		.expect("now always later than unix epoch; qed")
-		.as_secs()
+struct LocalDuty {
+	validation: Chain,
 }
 
 /// The Polkadot proposer logic.
-pub struct Proposer<C: PolkadotApi, R> {
-	parent_hash: HeaderHash,
-	parent_number: BlockNumber,
-	parent_id: C::CheckedBlockId,
+pub struct Proposer<C: PolkadotApi, R, P> {
 	client: Arc<C>,
+	collators: P,
+	delay: Shared<Sleep>,
+	dynamic_inclusion: DynamicInclusion,
+	local_duty: LocalDuty,
 	local_key: Arc<ed25519::Pair>,
+	parent_hash: HeaderHash,
+	parent_id: C::CheckedBlockId,
+	parent_number: BlockNumber,
+	random_seed: Hash,
+	router: R,
+	table: Arc<SharedTable>,
+	timer: Timer,
 	transaction_pool: Arc<Mutex<TransactionPool>>,
-	delay: Shared<Timeout>,
-	_table: Arc<SharedTable>,
-	_router: R,
 }
 
-impl<C: PolkadotApi, R: TableRouter> bft::Proposer for Proposer<C, R> {
+impl<C, R, P> bft::Proposer for Proposer<C, R, P>
+	where
+		C: PolkadotApi,
+		R: TableRouter,
+		P: Collators,
+{
 	type Error = Error;
-	type Create = Result<SubstrateBlock, Error>;
+	type Create = CreateProposal<C, R, P>;
 	type Evaluate = Box<Future<Item=bool, Error=Error>>;
 
 	fn propose(&self) -> Self::Create {
-		debug!(target: "bft", "proposing block on top of parent ({}, {:?})", self.parent_number, self.parent_hash);
+		const ATTEMPT_PROPOSE_EVERY: Duration = Duration::from_millis(100);
+
+		let initial_included = self.table.includable_count();
+		let enough_candidates = self.dynamic_inclusion.acceptable_in(
+			Instant::now(),
+			initial_included,
+		).unwrap_or_default();
+
+		CreateProposal {
+			parent_hash: self.parent_hash.clone(),
+			parent_number: self.parent_number.clone(),
+			parent_id: self.parent_id.clone(),
+			client: self.client.clone(),
+			transaction_pool: self.transaction_pool.clone(),
+			collation: CollationFetch::new(
+				self.local_duty.validation,
+				self.parent_id.clone(),
+				self.parent_hash.clone(),
+				self.collators.clone(),
+				self.client.clone()
+			),
+			table: self.table.clone(),
+			router: self.router.clone(),
+			timing: ProposalTiming {
+				timer: self.timer.clone(),
+				attempt_propose: self.timer.interval(ATTEMPT_PROPOSE_EVERY),
+				enough_candidates: self.timer.sleep(enough_candidates),
+				dynamic_inclusion: self.dynamic_inclusion.clone(),
+				minimum_delay: Some(self.delay.clone()),
```

Comment:
`enough_candidates` is actually more of a "enough candidates with the amount of candidates right now". Since this will change over time we need to periodically reschedule this future.